include/rnp/rnp2.h

```diff
@@ -0,0 +1,241 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*
+* Currently we use some definitions/enums from other others, including
+* - rnp_result
+* - pgp_hash_alg_t
+* - pgp_pubkey_alg_t
+*
+* It might be good to consolidate these definitions into one spot, or
+* else replicate them here in new enums and make this header freestanding.
+*/
+#include <rnp_def.h>
+#include <repgp_def.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+/*
+* Opaque structures for key types
+*/
+typedef struct rnp_publickey_st * rnp_public_key;
+typedef struct rnp_privatekey_st *rnp_private_key;
+
+// typedef struct rnp_keyring_st* rnp_keyring;
+
+/* Operations on public keys */
+
+/**
+* Load a public key
+* @param key a pointer, on success will be assigned to the new object
+* @param buf the serialized key (either binary or ASCII formats accepted)
+* @param buf_len length of buf in bytes
+*/
+rnp_result rnp_public_key_load(rnp_public_key *key, const uint8_t buf[], size_t buf_len);
+
+/**
+* @param key the public key
+* @param uid on success will be set to the UID of the key
+*/
+rnp_result rnp_public_key_get_uid(rnp_public_key key, char **uid);
+
+/**
+* @param key the public key
+* @param fprint on success will be set to the fingerprint of the key
+*/
+rnp_result rnp_public_key_get_fprint(rnp_public_key key, char **fprint);
+
+/* Operations on private keys */
+
+/**
+* Callback used for getting a passphrase.
+* @param app_ctx provided by application
+* @param key the key that is being decrypted
+* @param pass to which the callback should write the returned
+* passphrase, NULL terminated.
+* @param pass_len the size of pass buffer
+* @return 0 on success, or any other value to stop decryption.
+*/
+typedef int (*rnp_passphrase_cb)(void *          app_ctx,
+                                 rnp_private_key key,
+                                 char            buf[],
+                                 size_t          buf_len);
+
+/*
+* Load a private key into memory. Initially it will be encrypted,
+* assuming the input is encrypted.
+*/
+rnp_result rnp_private_key_load(rnp_private_key *key, const uint8_t buf[], size_t buf_len);
+
+rnp_result rnp_private_key_is_decrypted(rnp_private_key key);
+
+rnp_result rnp_private_key_decrypt(rnp_private_key key, rnp_passphrase_cb cb, void *app_ctx);
+
+/**
+* @param sub_alg may be 0 (NONE) to disable subkeys
+*/
+rnp_result rnp_generate_key(rnp_private_key *key,
+                            const char *     userid,
+                            pgp_pubkey_alg_t prim_alg,
+                            pgp_pubkey_alg_t sub_alg,
+                            pgp_hash_alg_t   signature_hash,
+                            uint32_t         expiration);
```

Comment:
What about have multiple functions like rnp_generate_key_ecdsa(), rnp_generate_key_rsa(keybits, subkeybits) and so on?

Reply:
My thought here was if sub_alg == NONE then it would not generate a subkey, otherwise it would. But, since a key can legitimately have more than one subkey, it would probably be better to have generate_key + add_subkey operations. (Plus I guess also remove_subkey, not totally sure if that's useful).

And yes this should provide a way to specify key params. I'm torn on a `generate_key` per algo type. One the one hand it makes things a little simpler in each API, but it causes a combinatoric explosion given we already suppport RSA, DSA, ECDSA, ECDH, EdDSA, ElGamal, and SM2. Yikes. I don't know what the best approach is here.

I wanted to avoid a union as is used in `rnp_keygen_crypto_params_t` but `union`s in FFI seem kind of messy/dangerous. Maybe this is fine after all.

Reply:
I think we should separate the API for generating subkeys since there could be more than one. Specifically, `generate_key` (main key and subkeys) and `add/remove_subkey` should be separated so we could set key params and preferences in a simple way.